### PR TITLE
Dataset features

### LINKED ISSUE #50: Util for programmatically creating dataset for processed files
**Is your feature request related to a problem? Please describe.**
Requirement is to organize processed products as datasets, but the experience of adding files from the UI isn't that great especially when the output is spread across different folders/several folders deep.

**Describe the solution you'd like**
Something like `new_dataset` that takes parameters 1) an output location (from a standard workflow) and 2) the type of data (e.g. level 2 vs level 3?) and creates a dataset with the expected files. 

**Additional context**
The Python client currently doesn't expose this yet, since datasets are fairly new. However, this should be possible with the lower-level REST API.

### LINKED ISSUE #104: Add script for generating dataset citations
A [DataCite citation](https://datacite.org/cite-your-data.html) looks like:
> Creator (PublicationYear). Title. Publisher. Identifier

Scientific Data [format](https://blogs.nature.com/scientificdata/2016/07/14/data-citations-at-scientific-data/):
> ### How do I format my data citations?
> Data citations at Scientific Data include the following four fields.

>    1. Creator/Authors(s)
      The creators of the dataset, which may be distinct from the authors of the Data Descriptor.
>    2. Repository Name 
      For DataCite DOIs this should align with the “Publisher” field in DataCite metadata.
>    3. Dataset identifier 
       This will be a DataCite DOI or an appropriate repository accession ID.
>    4. Dataset Publication Year
      The year the data were made publicly available.

> **_Example citation with a Digital Object Identifier_**
   Hibsh, D., Schori, H., Efroni, S. & Shefi, O. _Figshare_ https://dx.doi.org/10.6084/m9.figshare.1289242 (2015).

Currently, the dataset schema doesn't completely support generating such a citation format, e.g. we're missing publication year. However, once https://github.com/nf-osi/nf-metadata-dictionary/issues/200 is defined and we update metadata of current datasets, it would be possible to use the API to get Synapse metadata and spit out citation text in a format (e.g. shown above) that people want to use.

Note: This really should be a feature on Synapse that could be implemented with simple Javascript.

### LINKED ISSUE #105: Add utils for dataset collection manipulation
The Python client doesn't yet offer methods for programmatically manipulating dataset collections (see code for [table services](https://github.com/Sage-Bionetworks/synapsePythonClient/blob/develop/synapseclient/table.py)), which would be helpful for dataset management. In particular, certain workflows can be onerous to do through the dataset collections UI:

- Adding dataset items to a dataset collection can only be done through the UI. Depending on the projects under management, this can be a lot of inefficient clicking around and typing in Synapse ids in the UI.
- Updating datasets in a collection to the latest versions. While changing the Version selection dropdown for 1-2 datasets is OK, it is often necessary to update a large part of the collection or the entire collection of datasets to use "latest". 

These can be relatively easy to implement and can be swapped out later for the "official" implementation.

### PR DESCRIPTION
**Pull request by anngvu**
Dataset utilities:
- https://github.com/nf-osi/nfportalutils/issues/50 - Already in practical use at time of PR, revised initial implementation to address review comments, improve code reuse/flexibility.
- https://github.com/nf-osi/nfportalutils/issues/105 - To officially transition to dataset collections, we needed to implement dataset collection utils not currently in Python client.
- https://github.com/nf-osi/nfportalutils/issues/104 - Mostly for internal demonstration purposes only for the nf-specific datasets; this should ultimately be a portal feature like https://help.figshare.com/article/how-to-share-cite-or-embed-your-data (and better implemented in Javascript/Typescript).

### CODE REVIEW

File: R/datasets.R [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+#'}
+nf_star_salmon_datasets <- function(output_map, 
+                                    parent, 
+                                    verbose = TRUE, 
+                                    dry_run = TRUE) { 
+  
+  # Select the .sf and index files
+  dataset_items <- output_map[grepl(".sf$", output_name), output_id]
+  name <- "Gene Expression Quantification from RNA-seq"
+  dataset_items <- lapply(dataset_items, function(entity) list(entityId = entity, versionNumber = 1L))
```
Comment (allaway) [inline R/datasets.R]: Curious why this defaults to V1. Alternatively, can we assume the latest version is the "good" version if someone is making a dataset?

File: R/datasets.R [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+  pattern <- "vcf.gz(.tbi)?$"
+  WORKFLOW <- c("FreeBayes", "Mutect2", "Strelka", "DeepVariant")
+  datasets <- list()
+  for(i in WORKFLOW) {
+    dataset <- output_map[workflow == i & grepl(pattern, output_name)]
+    if(nrow(dataset)) {
+      if(verbose) glue::glue("Creating {i} dataset with {nrow(dataset)} files") 
+      
+      name <- glue::glue("{gvtype} Genomic Variants - {i} Pipeline")
+      dataset_items <- lapply(dataset$output_id, function(entity) list(entityId = entity, versionNumber = 1L))
```
Comment (allaway) [inline R/datasets.R]: Curious why this defaults to V1. Alternatively, can we assume the latest version is the "good" version if someone is making a dataset?

File: R/datasets.R [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+      "Somatic"  
+    } else if(grepl("GermlineVariantCalls", first(output_map$caller_path)))  {
+      "Germline"
+    } else {
+      stop("Could not assign either Germline or Somatic labels based on main output folder. 
+           Check whether folder contains mixed types or is not the right one.")
+    }
+  }
+  pattern <- "vcf.gz(.tbi)?$"
+  WORKFLOW <- c("FreeBayes", "Mutect2", "Strelka", "DeepVariant")
```
Comment (allaway) [inline R/datasets.R]: Should this be a parameter instead?

**Review by allaway — CHANGES_REQUESTED** [review]
Haven't tested the functions yet.

**Comment by anngvu** [comment]
@allaway There is one known test failure in https://github.com/nf-osi/nfportalutils/actions/runs/5449425127/jobs/9913626204?pr=74#step:6:1392 (`get_dependency_from_json_schema `) because we updated the data model. It's outside the scope/theme of this PR, so will fix later.

File: R/datasets.R [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+#' Update item versions to "latest" in a collection
+#' 
+#' Update an _existing_ collection so that all items or a subset of items reference their latest version.
+#' This should work for both datasets (collection of files) and dataset collections (collection of datasets).
+#' 
+#' @param collection_id Collection id.
+#' @param items Vector of dataset ids for which to update reference to latest version, 
+#' or "all" (default) to update all in the dataset collection.
+#' @export
+use_latest_in_collection <- function(collection_id, items = "all") {
```
Comment (allaway) [inline R/datasets.R:59]:
```suggestion
use_latest_in_collection <- function(collection_id, items = "all") {
```
This is great!!!

**Review by allaway — APPROVED** [review]
Looks good to me! I have not tested everything, but I have a dataset that I should be processing in the next week or so and will have the opportunity to do then and flag any issues.

**Pull request merged**